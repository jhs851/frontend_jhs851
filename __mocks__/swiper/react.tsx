import React from 'react'

const mock: React.FC = ({ children }) => <div>{children}</div>

export default mock
