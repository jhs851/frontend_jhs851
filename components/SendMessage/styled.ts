import styled from 'styled-components'
import { flexCenter } from 'styles/flex'

export const SendContainer = styled.form`
  display: flex;
  padding: 10px 16px 20px;
  position: absolute;
  left: 0;
  right: 0;
  bottom: 0;
  background-color: ${({ theme }) => theme.paleGrey};
`

export const TextareaContainer = styled.div`
  flex: 1;
  height: 50px;
  margin-right: 12px;
  padding: 17px 16px 16px;
  border-radius: 25px;
  box-shadow: 0 2px 4px 0 ${({ theme }) => theme.black10};
  background-color: ${({ theme }) => theme.background};
`

export const SubmitContainer = styled.button`
  width: 50px;
  height: 50px;
  border-radius: 50%;
  background-color: ${({ theme }) => theme.primary};
  transition: opacity 300ms;
  ${flexCenter}

  :disabled {
    opacity: 0.4;
  }
`
