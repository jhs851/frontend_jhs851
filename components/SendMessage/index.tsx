import React from 'react'
import { SendContainer, SubmitContainer, TextareaContainer } from './styled'
import Textarea from 'components/Textarea'
import Image from 'components/Image'
import { useForm } from 'react-hook-form'

type P = {
  onSubmit: (input: MessageFormInput) => Promise<void>
}

export type MessageFormInput = {
  message: string
  asset: string
}

const SendMessage: React.FC<P> = ({ onSubmit }) => {
  const {
    register,
    handleSubmit,
    reset,
    formState: { isValid },
  } = useForm<MessageFormInput>({
    mode: 'onChange',
    defaultValues: {
      message: '',
      asset: '',
    },
  })
  const submit = async (input: MessageFormInput) => {
    reset({ message: '' })
    await onSubmit(input)
  }

  return (
    <SendContainer onSubmit={handleSubmit(submit)}>
      <TextareaContainer>
        <Textarea
          placeholder="메시지를 입력하세요.."
          rows={1}
          {...register('message', {
            validate: val => Boolean(val.trim()),
          })}
        />
      </TextareaContainer>

      <SubmitContainer disabled={!isValid} type="submit">
        <Image src="/icons/send.svg" width={26} height={18} alt="전송" />
      </SubmitContainer>
    </SendContainer>
  )
}

export default SendMessage
