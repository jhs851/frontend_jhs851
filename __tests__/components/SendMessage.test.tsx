import { fireEvent, render, waitFor } from '@testing-library/react'
import SendMessage from 'components/SendMessage'

describe('SendMessage 컴포넌트', () => {
  const onSubmit = jest.fn()
  const rendered = () => render(<SendMessage onSubmit={onSubmit} />)

  it('렌더링이 올바르게 된다', () => {
    expect(rendered().container).toMatchSnapshot()
  })

  it('메세지 값이 없으면 submit 버튼은 꺼져있다', () => {
    const { getByRole } = rendered()

    expect(getByRole('button')).toBeDisabled()
  })

  it('메세지를 입력하고 submit 버튼을 누르면 onSubmit이 실행된다', async () => {
    const { getByPlaceholderText, getByRole } = rendered()
    const value = '메세지'
    const textarea = getByPlaceholderText(
      '메시지를 입력하세요..',
    ) as HTMLTextAreaElement

    fireEvent.change(textarea, {
      target: { value },
    })
    expect(textarea.value).toBe(value)
    await waitFor(() => {
      fireEvent.submit(getByRole('button'))
    })
    expect(onSubmit).toHaveBeenCalled()
  })
})
