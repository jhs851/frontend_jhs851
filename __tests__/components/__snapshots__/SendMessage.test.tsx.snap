// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SendMessage 컴포넌트 렌더링이 올바르게 된다 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 10px 16px 20px;
  position: absolute;
  left: 0;
  right: 0;
  bottom: 0;
}

.c1 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  height: 50px;
  margin-right: 12px;
  padding: 17px 16px 16px;
  border-radius: 25px;
  box-shadow: 0 2px 4px 0;
}

.c3 {
  width: 50px;
  height: 50px;
  border-radius: 50%;
  -webkit-transition: opacity 300ms;
  transition: opacity 300ms;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c3:disabled {
  opacity: 0.4;
}

.c2 {
  font-family: AppleSDGothicNeo,sans-serif;
  font-size: 14px;
  -webkit-letter-spacing: -0.12px;
  -moz-letter-spacing: -0.12px;
  -ms-letter-spacing: -0.12px;
  letter-spacing: -0.12px;
  width: 100%;
  resize: none;
  -ms-overflow-style: none;
  -webkit-scrollbar-width: none;
  -moz-scrollbar-width: none;
  -ms-scrollbar-width: none;
  scrollbar-width: none;
}

.c2 {
  font-family: AppleSDGothicNeo,sans-serif;
  font-size: 14px;
  -webkit-letter-spacing: -0.12px;
  -moz-letter-spacing: -0.12px;
  -ms-letter-spacing: -0.12px;
  letter-spacing: -0.12px;
  width: 100%;
  resize: none;
  -ms-overflow-style: none;
  -webkit-scrollbar-width: none;
  -moz-scrollbar-width: none;
  -ms-scrollbar-width: none;
  scrollbar-width: none;
}

.c2 {
  font-family: AppleSDGothicNeo,sans-serif;
  font-size: 14px;
  -webkit-letter-spacing: -0.12px;
  -moz-letter-spacing: -0.12px;
  -ms-letter-spacing: -0.12px;
  letter-spacing: -0.12px;
  width: 100%;
  resize: none;
  -ms-overflow-style: none;
  -webkit-scrollbar-width: none;
  -moz-scrollbar-width: none;
  -ms-scrollbar-width: none;
  scrollbar-width: none;
}

.c2 {
  font-family: AppleSDGothicNeo,sans-serif;
  font-size: 14px;
  -webkit-letter-spacing: -0.12px;
  -moz-letter-spacing: -0.12px;
  -ms-letter-spacing: -0.12px;
  letter-spacing: -0.12px;
  width: 100%;
  resize: none;
  -ms-overflow-style: none;
  -webkit-scrollbar-width: none;
  -moz-scrollbar-width: none;
  -ms-scrollbar-width: none;
  scrollbar-width: none;
}

.c2::-webkit-scrollbar {
  display: none;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<div>
  <form
    class="c0"
  >
    <div
      class="c1"
    >
      <textarea
        class="c2"
        name="message"
        placeholder="메시지를 입력하세요.."
        rows="1"
      />
    </div>
    <button
      class="c3"
      disabled=""
      type="submit"
    >
      <div
        class="c4"
      />
    </button>
  </form>
</div>
`;
